pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1106,12 +1106,324 @@ func (s *sqlEntityServer) List(ctx context.Context, r *entity.EntityListRequest)
 	return rsp, err
 }
 
-func (s *sqlEntityServer) Watch(*entity.EntityWatchRequest, entity.EntityStore_WatchServer) error {
+func (s *sqlEntityServer) Watch(r *entity.EntityWatchRequest, w entity.EntityStore_WatchServer) error {
 	if err := s.Init(); err != nil {
 		return err
 	}
 
-	return fmt.Errorf("unimplemented")
+	user, err := appcontext.User(w.Context())
+	if err != nil {
+		return err
+	}
+	if user == nil {
+		return fmt.Errorf("missing user in context")
+	}
+
+	if r.Since == 0 {
+		err = s.watchInit(w.Context(), r, w)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		err := w.Context().Err()
+		if err != nil {
+			s.log.Debug("watch context cancelled", "err", err)
+			break
+		}
+
+		err = s.watch(w.Context(), r, w)
+		if err != nil {
+			s.log.Error("watch error", "err", err)
+			return err
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+
+	return nil
+}
+
+// watchInit is a helper function to send the initial set of entities to the client
+func (s *sqlEntityServer) watchInit(ctx context.Context, r *entity.EntityWatchRequest, w entity.EntityStore_WatchServer) error {
```

Comment:
In theory yes – I was commenting on the current design. But even if we do share watches, the logic behind them is still expected to be the same – we just list the changes between RV versions X and Y, batch them and push them onto the wire – there isn't a difference between the init and consecutive batches.

Reply:
> the logic behind them is still expected to be the same – we just list the changes between RV versions X and Y

good point

Reply:
not sure if this might help identify expected behavior, but there are [watch tests here](https://github.com/kubernetes/kubernetes/blob/73f19e4c0162a2ee711c254b2af226b4e135d0e3/staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go#L48)

Reply:
@DanCech thanks for the link! Turns out I was misremembering how it was supposed to work and the current implementation is correct!

I think that we still need to refactor watchInit / watch to share things like filters / batch sizes though – currently there's a lot of duplicated logic.

Reply:
yeah, all this code is just hacked together enough to get it working, long term the intent is that watchinit would continue to work more or less as it does now (it's really just doing the list part of list+watch inline instead of the client getting the list then starting the watch from the returned revision).

For the watch phase I'm going to take a stab at a broadcaster model so that we can have one poller/log watcher supporting multiple watch clients.

Reply:
SGTM! FYI here's [Kine's broadcaster implementation](https://github.com/k3s-io/kine/blob/37736729c40f1e7ad18521d1eb5c2a13e314ebde/pkg/broadcaster/broadcaster.go) and [some details on its usage](https://github.com/k3s-io/kine/blob/37736729c40f1e7ad18521d1eb5c2a13e314ebde/docs/flow.md#flow-details).